File: internal/rules/consistent_return/consistent_return_test.go
package consistent_return

import (
	"testing"

	"github.com/typescript-eslint/rslint/internal/rule_tester"
	"github.com/typescript-eslint/rslint/internal/rules/fixtures"
)

func TestConsistentReturnRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &ConsistentReturnRule, []rule_tester.ValidTestCase{
		// Base rule cases
		{Code: `
function foo() {
  return;
}
		`},
		{Code: `
const foo = (flag: boolean) => {
  if (flag) return true;
  return false;
};
		`},
		{Code: `
class A {
  foo() {
    if (a) return true;
    return false;
  }
}
		`},
		{
			Code: `
const foo = (flag: boolean) => {
  if (flag) return;
  else return undefined;
};
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": true,
			},
		},
		// Void return type cases
		{Code: `
declare function bar(): void;
function foo(flag: boolean): void {
  if (flag) {
    return bar();
  }
  return;
}
		`},
		{Code: `
declare function bar(): void;
const foo = (flag: boolean): void => {
  if (flag) {
    return;
  }
  return bar();
};
		`},
		{Code: `
function foo(flag?: boolean): number | void {
  if (flag) {
    return 42;
  }
  return;
}
		`},
		{Code: `
function foo(): boolean;
function foo(flag: boolean): void;
function foo(flag?: boolean): boolean | void {
  if (flag) {
    return;
  }
  return true;
}
		`},
		{Code: `
class Foo {
  baz(): void {}
  bar(flag: boolean): void {
    if (flag) return baz();
    return;
  }
}
		`},
		{Code: `
declare function bar(): void;
function foo(flag: boolean): void {
  function fn(): string {
    return '1';
  }
  if (flag) {
    return bar();
  }
  return;
}
		`},
		{Code: `
class Foo {
  foo(flag: boolean): void {
    const bar = (): void => {
      if (flag) return;
      return this.foo();
    };
    if (flag) {
      return this.bar();
    }
    return;
  }
}
		`},
		// Async cases
		{Code: `
declare function bar(): void;
async function foo(flag?: boolean): Promise<void> {
  if (flag) {
    return bar();
  }
  return;
}
		`},
		{Code: `
declare function bar(): Promise<void>;
async function foo(flag?: boolean): Promise<ReturnType<typeof bar>> {
  if (flag) {
    return bar();
  }
  return;
}
		`},
		{Code: `
async function foo(flag?: boolean): Promise<Promise<void | undefined>> {
  if (flag) {
    return undefined;
  }
  return;
}
		`},
		{Code: `
type PromiseVoidNumber = Promise<void | number>;
async function foo(flag?: boolean): PromiseVoidNumber {
  if (flag) {
    return 42;
  }
  return;
}
		`},
		{Code: `
class Foo {
  baz(): void {}
  async bar(flag: boolean): Promise<void> {
    if (flag) return baz();
    return;
  }
}
		`},
		// treatUndefinedAsUnspecified option cases
		{
			Code: `
declare const undef: undefined;
function foo(flag: boolean) {
  if (flag) {
    return undef;
  }
  return 'foo';
}
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": false,
			},
		},
		{
			Code: `
function foo(flag: boolean): undefined {
  if (flag) {
    return undefined;
  }
  return;
}
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": true,
			},
		},
		{
			Code: `
declare const undef: undefined;
function foo(flag: boolean): undefined {
  if (flag) {
    return undef;
  }
  return;
}
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": true,
			},
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code: `
function foo(flag: boolean): any {
  if (flag) return true;
  else return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "missingReturnValue",
					Line:      4,
					Column:    8,
					EndLine:   4,
					EndColumn: 15,
				},
			},
		},
		{
			Code: `
function bar(): undefined {}
function foo(flag: boolean): undefined {
  if (flag) return bar();
  return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "missingReturnValue",
					Line:      5,
					Column:    3,
					EndLine:   5,
					EndColumn: 10,
				},
			},
		},
		{
			Code: `
declare function foo(): void;
function bar(flag: boolean): undefined {
  function baz(): undefined {
    if (flag) return;
    return undefined;
  }
  if (flag) return baz();
  return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      6,
					Column:    12,
					EndLine:   6,
					EndColumn: 21,
				},
				{
					MessageId: "missingReturnValue",
					Line:      9,
					Column:    3,
					EndLine:   9,
					EndColumn: 10,
				},
			},
		},
		{
			Code: `
function foo(flag: boolean): Promise<void> {
  if (flag) return Promise.resolve(void 0);
  else return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "missingReturnValue",
					Line:      4,
					Column:    8,
					EndLine:   4,
					EndColumn: 15,
				},
			},
		},
		{
			Code: `
async function foo(flag: boolean): Promise<string> {
  if (flag) return;
  else return 'value';
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      4,
					Column:    15,
					EndLine:   4,
					EndColumn: 22,
				},
			},
		},
		{
			Code: `
async function foo(flag: boolean): Promise<string | undefined> {
  if (flag) return 'value';
  else return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "missingReturnValue",
					Line:      4,
					Column:    8,
					EndLine:   4,
					EndColumn: 15,
				},
			},
		},
		{
			Code: `
async function foo(flag: boolean) {
  if (flag) return;
  return 1;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      4,
					Column:    10,
					EndLine:   4,
					EndColumn: 11,
				},
			},
		},
		{
			Code: `
function foo(flag: boolean): Promise<string | undefined> {
  if (flag) return;
  else return 'value';
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      4,
					Column:    15,
					EndLine:   4,
					EndColumn: 22,
				},
			},
		},
		{
			Code: `
declare function bar(): Promise<void>;
function foo(flag?: boolean): Promise<void> {
  if (flag) {
    return bar();
  }
  return;
}
			`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "missingReturnValue",
					Line:      7,
					Column:    3,
					EndLine:   7,
					EndColumn: 10,
				},
			},
		},
		// treatUndefinedAsUnspecified option cases
		{
			Code: `
function foo(flag: boolean): undefined | boolean {
  if (flag) {
    return undefined;
  }
  return true;
}
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": true,
			},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      6,
					Column:    10,
					EndLine:   6,
					EndColumn: 14,
				},
			},
		},
		{
			Code: `
declare const undefOrNum: undefined | number;
function foo(flag: boolean) {
  if (flag) {
    return;
  }
  return undefOrNum;
}
			`,
			Options: map[string]interface{}{
				"treatUndefinedAsUnspecified": true,
			},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "unexpectedReturnValue",
					Line:      7,
					Column:    10,
					EndLine:   7,
					EndColumn: 20,
				},
			},
		},
	})
}

File: packages/rslint-test-tools/tests/eslint/rules/consistent-return.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('consistent-return', {
  valid: [
    'function foo() { return; }',
    'function foo() { if (true) return; }',
    'function foo() { if (true) return; else return; }',
    'function foo() { if (true) return true; else return false; }',
    'f(function() { return; })',
    'f(function() { if (true) return; })',
    'f(function() { if (true) return; else return; })',
    'f(function() { if (true) return true; else return false; })',
    'function foo() { function bar() { return true; } return; }',
    'function foo() { function bar() { return; } return false; }',
    'function Foo() { if (!(this instanceof Foo)) return new Foo(); }',
    'function foo() { if (true) return 5; else return undefined; }',
    'function foo() { if (true) return 5; else return void 0; }',
    {
      code: 'function foo() { if (true) return; else return undefined; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    {
      code: 'function foo() { if (true) return; else return void 0; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    {
      code: 'function foo() { if (true) return undefined; else return; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    {
      code: 'function foo() { if (true) return undefined; else return void 0; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    {
      code: 'function foo() { if (true) return void 0; else return; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    {
      code: 'function foo() { if (true) return void 0; else return undefined; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
    },
    'var x = () => {  return {}; };',
    'if (true) { return 1; } return 0;',

    // https://github.com/eslint/eslint/issues/7790
    'class Foo { constructor() { if (true) return foo; } }',
    'var Foo = class { constructor() { if (true) return foo; } }',
  ],
  invalid: [
    {
      code: 'function foo() { if (true) return true; else return; }',
      errors: [{ messageId: 'missingReturnValue', column: 46 }],
    },
    {
      code: 'var foo = () => { if (true) return true; else return; }',
      errors: [{ messageId: 'missingReturnValue', column: 47 }],
    },
    {
      code: 'function foo() { if (true) return; else return false; }',
      errors: [{ messageId: 'unexpectedReturnValue', column: 41 }],
    },
    {
      code: 'f(function() { if (true) return true; else return; })',
      errors: [{ messageId: 'missingReturnValue', column: 44 }],
    },
    {
      code: 'f(function() { if (true) return; else return false; })',
      errors: [{ messageId: 'unexpectedReturnValue', column: 39 }],
    },
    {
      code: 'f(a => { if (true) return; else return false; })',
      errors: [{ messageId: 'unexpectedReturnValue', column: 33 }],
    },
    {
      code: 'function foo() { if (true) return true; return undefined; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
      errors: [{ messageId: 'missingReturnValue', column: 41 }],
    },
    {
      code: 'function foo() { if (true) return true; return void 0; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
      errors: [{ messageId: 'missingReturnValue', column: 41 }],
    },
    {
      code: 'function foo() { if (true) return undefined; return true; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
      errors: [{ messageId: 'unexpectedReturnValue', column: 46 }],
    },
    {
      code: 'function foo() { if (true) return void 0; return true; }',
      options: [{ treatUndefinedAsUnspecified: true }] as any,
      errors: [{ messageId: 'unexpectedReturnValue', column: 43 }],
    },
    {
      code: 'if (true) { return 1; } return;',
      errors: [{ messageId: 'missingReturnValue', column: 25 }],
    },
    {
      code: 'function foo() { if (a) return true; }',
      errors: [{ messageId: 'missingReturn', column: 10 }],
    },
    {
      code: 'function _foo() { if (a) return true; }',
      errors: [{ messageId: 'missingReturn', column: 10 }],
    },
    {
      code: 'f(function foo() { if (a) return true; });',
      errors: [{ messageId: 'missingReturn', column: 12 }],
    },
    {
      code: 'f(function() { if (a) return true; });',
      errors: [{ messageId: 'missingReturn', column: 3 }],
    },
    {
      code: 'f(() => { if (a) return true; });',
      errors: [{ messageId: 'missingReturn', column: 6 }],
    },
    {
      code: 'var obj = {foo() { if (a) return true; }};',
      errors: [{ messageId: 'missingReturn', column: 12 }],
    },
    {
      code: 'class A {foo() { if (a) return true; }};',
      errors: [{ messageId: 'missingReturn', column: 10 }],
    },
    {
      code: 'if (a) return true;',
      errors: [{ messageId: 'missingReturn', column: 1 }],
    },
    {
      code: 'class A { CapitalizedFunction() { if (a) return true; } }',
      errors: [{ messageId: 'missingReturn', column: 11 }],
    },
    {
      code: '({ constructor() { if (a) return true; } });',
      errors: [{ messageId: 'missingReturn', column: 4 }],
    },
  ],
});
